test/integration/quic_http_integration_test.cc

```diff
@@ -347,10 +345,118 @@ class QuicHttpIntegrationTest : public HttpIntegrationTest,
   bool validation_failure_on_path_response_{false};
 };
 
+class QuicHttpIntegrationTest : public QuicHttpIntegrationTestBase,
+                                public testing::TestWithParam<Network::Address::IpVersion> {
+public:
+  QuicHttpIntegrationTest()
+      : QuicHttpIntegrationTestBase(GetParam(), ConfigHelper::quicHttpProxyConfig()) {}
+};
+
+class QuicHttpMultiAddressesIntegrationTest
+    : public QuicHttpIntegrationTestBase,
+      public testing::TestWithParam<Network::Address::IpVersion> {
+public:
+  QuicHttpMultiAddressesIntegrationTest()
+      : QuicHttpIntegrationTestBase(GetParam(), ConfigHelper::quicHttpProxyConfig(true)) {}
+
+  void testMultipleQuicConnections() {
+    // Enabling SO_REUSEPORT with 8 workers. Unfortunately this setting makes the test rarely flaky
+    // if it is configured to run with --runs_per_test=N where N > 1 but without --jobs=1.
+    setConcurrency(8);
+    initialize();
+    std::vector<std::string> addresses({"address1", "address2"});
+    registerTestServerPorts(addresses);
+    std::vector<IntegrationCodecClientPtr> codec_clients1;
+    std::vector<IntegrationCodecClientPtr> codec_clients2;
+    for (size_t i = 1; i <= concurrency_; ++i) {
+      // The BPF filter and ActiveQuicListener::destination() look at the 1st word of connection id
+      // in the packet header. And currently all QUIC versions support >= 8 bytes connection id. So
+      // create connections with the first 4 bytes of connection id different from each
+      // other so they should be evenly distributed.
+      designated_connection_ids_.push_back(quic::test::TestConnectionId(i << 32));
+      // TODO(sunjayBhatia,wrowe): deserialize this, establishing all connections in parallel
+      // (Expected to save ~14s each across 6 tests on Windows)
+      codec_clients1.push_back(makeHttpConnection(lookupPort("address1")));
+      designated_connection_ids_.push_back(quic::test::TestConnectionId(i << 32));
```

Comment:
Here codec_clients1[i] and codec_clients2[i] use the same connection ID and they connect to the same listener (on different ports). I thought this would be a connection ID collision, but the test seems to work. Maybe I missed some obvious facts, but can you comment more on what this test is testing and how it works?

Reply:
When using multiple addresses in a single listener, we will have single ListenerImpl for it. But actually, we create multiple separated ActiveQuicListener for each address.

For the same connection id's `codec_clients1` and `codec_clients2` will go to the different ActiveQuicListener instance, so there is no collision.

The multiple addresses didn't touch QUIC direclty accept the change of UdpWorkerRouter https://github.com/envoyproxy/envoy/pull/21940

That PR ensures the udp/quic request going to the correct worker and correct active listener.

If there is a bug that let the connection with the same id but with a different destination address go to the same listener, then this test will fail (fail with the connection ID collision as you said).

Let me add a comment here to explain this. But also let me know if I didn't understand quic code correctly.

Thanks for the review!

Reply:
Okay, I thought the same ActiveQuicListener would listen on both addresses, but in fact they are still 1:1 mapping. So this test is verifying that both listeners are dispatching quic packets according to connection IDs correctly. It makes sense! Thanks for explanation!

And it worth pointing out that such implementation further disassociate supporting QUIC server preferred address with what Greg proposed here: https://github.com/envoyproxy/envoy/issues/11184#issuecomment-679214885, which requires the same listener to listen on all the configured addresses. This might not be a issue, but worth clarifying on the PR about that.

Reply:
> And it worth pointing out that such implementation further disassociate supporting QUIC server preferred address with what Greg proposed here: [#11184 (comment)](https://github.com/envoyproxy/envoy/issues/11184#issuecomment-679214885), which requires the same listener to listen on all the configured addresses. This might not be a issue, but worth clarifying on the PR about that.

Thanks for pointing this out, yes, that case isn't the scope of this implementation. But I will think about it in the future steps, it seems like can be done with some change for UdpWorkerRouter, but anyway I will think about it more in the next steps.

Reply:
@danzh2010 I add a comment for that, PTAL again, thanks in advance!